File: Example 9 JNA/src/Example9.java
// Projeto : JNA Example #9
// Classe: Example9.java
// Função : Receive an Array of Doubles from C
// Autor(es) : Hugo Dionizio Santos
// Data : Seg Mai 11 11:05:54 BRT 2015

import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.ptr.PointerByReference;		
import com.sun.jna.Pointer;
import com.sun.jna.Structure;
import com.sun.jna.Structure.ByReference;
import com.sun.jna.ptr.IntByReference;

public class Example9 {
	public interface CLibrary extends Library {
		public double example9_getDoubleArray(PointerByReference valsRef, IntByReference numValsRef);
		public void example9_cleanup(Pointer p);
}

	static public void main(String argv[]) {
		CLibrary clib;
		PointerByReference ex9ValsRefPtr;

		try {
			clib = (CLibrary)Native.loadLibrary("./libexample9.so", CLibrary.class);
			// Receive an Array of Doubles from C
			ex9ValsRefPtr = new PointerByReference();
		}
		catch (UnsatisfiedLinkError e) {
			clib = (CLibrary)Native.loadLibrary("./bin/libexample9.so", CLibrary.class);
			// Receive an Array of Doubles from C
			ex9ValsRefPtr = new PointerByReference();
		}

		IntByReference ex9NumValsRef = new IntByReference();

		// call the C function
		clib.example9_getDoubleArray(ex9ValsRefPtr, ex9NumValsRef);

		// extract the number of values returned
		int ex9NumVals = ex9NumValsRef.getValue();
		System.out.println("example 9: retreived " + ex9NumVals + " values:");

		// make sure the C function returned some values
		if (0 < ex9NumVals) {
			// extract the pointed-to pointer
			Pointer ex9pVals = ex9ValsRefPtr.getValue();
			// look at each value
			double ex9total = 0.0;
			for (int ex9Loop=0; ex9Loop<ex9NumVals; ex9Loop++) {
				double ex9val = ex9pVals.getDouble(ex9Loop * Native.getNativeSize(Double.TYPE));
				ex9total += ex9val;
			}
			System.out.println("\ttotal: " + ex9total);
	
			// call C code to clean up memory allocated in C
			System.out.println("\t(example 9: cleanup)");
			clib.example9_cleanup(ex9pVals);
		}		
	}
}

// Fim da classe

File: Example 9 JNA/src/example9.c
// Projeto : JNA Example #9
// Algoritmo: example9.c
// Função : Receive an Array of Doubles from C
// Autor(es) : Hugo Dionizio Santos (adaptações e correções), Ethan Shayne in http://www.eshayne.com/jnaex
// Data : Seg Mai 11 11:05:54 BRT 2015

#include <stdlib.h>

// Seção de Variáveis Globais

// Interface do algoritmo

// Função Principal
//	Seção de Variáveis Locais
//	Seção de Comandos

// Seção de Funções
void example9_getDoubleArray(double** ppVals, int* pNumVals)
{
	int loop = 0;
	*pNumVals = 100;
	*ppVals = (double*)malloc(sizeof(double) * *pNumVals);
	memset(*ppVals, 0, sizeof(double) * *pNumVals);
	for (loop=0; loop<*pNumVals; loop++)
	{
		// populate the array with junk data (just for the sake of the example)
		(*ppVals)[loop] = ((double)loop + 100) / 100;
	}
}

void example9_cleanup(double* pVals)
{
	free(pVals);
}
					
				


// Fim do algoritmo
